File: Baekjoon/Python/DisjointSet/Q1717.py
# 집합의 표현 / 서로소 집합(유니온 파인드)
import sys

def find(x):
    if parent[x] != x:
        parent[x] = find(parent[x])
    return parent[x]

def union(a, b):
    a = find(a)
    b = find(b) 

    if a < b:
        parent[b] = a
    else:
        parent[a] = b

sys.setrecursionlimit(100000)
input = sys.stdin.readline
n, m = map(int, input().split())
parent = [i for i in range(n+1)]

for _ in range(m):
    cmd, a, b = map(int, input().split())
    if cmd == 0:
        union(a, b)
    elif cmd == 1:
        if find(a) == find(b):
            print('YES')
        else:
            print('NO')

File: Baekjoon/Swift/BOJ_PS/BOJ_PS/DisJoint/Q1717.swift
//
//  main.swift
//  BOJ_PS
//
//  Created by 김태훈 on 2021/05/04.
//  집합의 표현 / 서로소(분리) 집합

let nm = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = nm[0]
let m = nm[1]
var parent = [Int](repeating: 0, count: n+1)

for i in 1 ..< n+1 {
    parent[i] = i
}

func find(_ x: Int) -> Int {
    if parent[x] != x {
        parent[x] = find(parent[x])
    }
    return parent[x]
}

func union(_ a: Int, _ b: Int) {
    let a = find(a)
    let b = find(b)
    
    if a < b {
        parent[b] = a
    } else {
        parent[a] = b
    }
}

for _ in 0 ..< m {
    let input = readLine()!.split(separator: " ").map { Int(String($0))! }
    
    if input[0] == 0 {
        union(input[1], input[2])
    } else {
        if find(input[1]) == find(input[2]) {
            print("YES")
        } else{
            print("NO")
        }
    }
}
